File: java/src/main/java/com/example/cote/PGMRS/주식_가격.java
package com.example.cote.PGMRS;

import java.util.*;
import java.util.stream.*;

public class 주식_가격 {

  private static final int MAX_PRICES_LENGTH = 100000;
  private static final int MAX_PRICE_VALUE = 10000;
  private static final int TEST_CASES = 10;
  private static Random random = new Random(43);

  private static int[] getPrices() {

    int n = random.nextInt(MAX_PRICES_LENGTH - 1) + 1;

    return IntStream.range(0, n).map(i -> random.nextInt(MAX_PRICE_VALUE - 1) + 1).toArray();
  }

  private static int[] solve(int[] prices) {
    int[] result = new int[prices.length];
    Stack<Integer> stack = new Stack<>();

    for (int i = 0; i < prices.length; i++) {

      while (!stack.isEmpty()) {
        int top = stack.peek();
        if (prices[top] <= prices[i]) {
          break;
        }
        int start = stack.pop();
        result[start] = i - start;
      }

      stack.push(i);
    }

    while (!stack.isEmpty()) {
      int start = stack.pop();
      result[start] = prices.length - 1 - start;
    }

    return result;
  }

  public static void main(String... args) {

    int[] sample_prices = new int[] {1, 2, 3, 2, 3};
    System.out.println(Arrays.toString(solve(sample_prices)));

    for (int i = 0; i < TEST_CASES; i++) {
      int[] prices = getPrices();
      System.out.println(Arrays.toString(solve(prices)));
    }
  }
}

File: kotlin/src/main/kotlin/com/example/cote/PGMRS/주식_가격.kt
package com.example.cote.PGMRS

import java.util.*
import kotlin.random.*

object 주식_가격 {

    val TEST_CASES = 10
    val MAX_PRICES_LENGTH = 10000
    val MAX_PRICES_VALUE = 10000

    val random = Random(43)

    // 가격이 떨어지지 않은 기간을 남은 배열을 반환
    fun solve(prices: List<Int>): List<Int> {
        val ret = MutableList<Int>(prices.size) { 0 }
        val stack = Stack<Int>()

        for ((i, price) in prices.withIndex()) {

            while (!stack.isEmpty()) {
                val top = stack.peek()
                if (prices[top] <= price) {
                    break
                }
                ret[top] = i - top
                stack.pop()
            }
            stack.push(i)
        }

        while (!stack.isEmpty()) {
            val top = stack.pop()
            ret[top] = prices.size - 1 - top
        }

        return ret
    }

    fun main() {
        repeat(TEST_CASES) {
            val prices =
                List(random.nextInt(MAX_PRICES_LENGTH)) {
                    random.nextInt(MAX_PRICES_VALUE)
                }
            println(solve(prices))
        }
    }
}

fun main() {
    주식_가격.main()
}
